README.md
# JGMOD
 


src/history.txt
Well. Here is the progress of JGMOD library.


December 1997
-------------
24  Got 2 MOD documentation from oulu. Neither give a very good description
    about the MOD format but with both combined, just enough.

25  Merry Christmas !! Decided to make a mod library and I called it JGMOD.
    You might be wondering what JG means. My actual name is Guan Foo Wah.
    Guan is my surname and people call me Jeffery, or Jeff. So JG stands for
    Jeffery Guan. I could call it GFWMOD but it really sounds terrible to me.

    Started making the library. The birth of this library !!! Able to load
    some of the MOD 31 instruments header (song name and all the 31
    instruments info). 

26  Revamped in the loading header function. It was untidy. Now it uses
    Allegro's pack_stuff routines. Able to load the MOD pattern for 4 channels
    mod.

27  Finally figured out how to load X channel mod. It is really easy. Some
    other damm documentation confuses me with it.  Able to load all the 31
    instruments sample. Having trouble to play those sample using Allegro's
    play_sample() function. Sounds too terrible.

28  Tried minus 127 with all of those 31 instruments samples and hell it
    works !! play_sample() have no problem playing those samples now. That
    was a lucky try. Made an interrupt to play those mod files assuming the
    music is at tempo 6 and bmp 125. Since it uses play_sample(), a new
    sample played will not cancel any sample on the same channel. Echos and
    feedback will be heard. Yuck !

29  Started making this history to whovever is interested.
    Now it uses Allegro's lower level digital sample routines. No more
    feedback and echos. YEAAA... Plays samples in loop mode if required.

30  Now plays the sample at the correct volume. Pattern Break command (13)
    is now supported. Able to play music at different tempo and bmp(15).

31  Supports Fine Volume Slide Up and Fine Volume Slide Down (14-10
    and 14-11).


January 1998
------------
3   Supports set sample offset(9).

5   Fixed Fine Volume Slide Up and Fine Volume Slide down bug

7   Changed the way to handle bpm by changing the interrupt rate
    when needed.

8   Fixed a pattern break bug. Supports MOD 15 instruments now. 

9   Volume slide (10) supported

10  Read the S3M format. Could not understand much. Change to NTSC clock rate
    cause S3M c2spd is in that rate. Implemented the finetune effect (14-5).

11  Implemented the period slide up (pormento up, 1) and slide
    down(promento down, 2) effect. Slide to note(3) is also supported but
    buggy.

12  Fixed slide to note bug, but a bug still exists.

26  Fixed another slide to note bug. I think it is free of bug now. Did a
    makefile for this library.

28  Supports Retrigger Sample (14-9) and Cut Sample (14-12).

29  Delay Sample (14-13) and Pormento to note+volume slide suported (5).
    Vibrato+volume slide is also supported but without the vibrato.
    Make prev_pattern() and next_pattern() to skip pattern.

30  Now the song can play in a loop mode. Fixed a bug in the skip pattern
    functions.  


Febuary 1998
------------
13  Opps. Haven't update this library for a long time. Did stop_mod function
    to stop the music. Updated the mod loader to support up to 32 channels
    but not player.

14  Did a loader for XM partially.

16  The player now supports up to 32 channels and the number of channels
    reserved for mod can be specified instead of the default 24 channels.

19  Changed how the mod interrupt works. Global commands will not be skiped
    if the no of channels reserved for the player is less than no of
    channels used by the song. Did a readme.txt. I am going to upload this to
    my web page tomorrow.

20  Library uploaded. 

23  Did the vibrato effect (4). Yea... !!!!

24  Added some preprocessor commands to jgmod.h. Did a cpp program to test
    with this library. It compiled fine. No errors at all. But when I ran the
    program, there is no sound at all. I have no clue at all why this
    is happening. I don't know much about cpp.

25  Did set pan effect (8) and 16 pos panning effect (14-8). This is a
    non-standard protracker effect. So, could not test my codes. Not sure if
    it really works. Also, did is_mod_playing() function to test the mod
    is playing anything.

27  Did pause_mod(), resume_mod() and is_mod_paused() function. Thought
    I found a bug in mod_interrupt() function in player.c. The player exits
    with SIGSEGV when a music playing in unloop mode ends without windows
    running. It has no problem when windows is running. After 2 hours of
    wondering, found the ass code. Jgmod.c is the problem. Jgmod.c is
    accessing 'of' pointer even it is pointing to NULL. Fixed the problem. If
    you are reading this, my advice. If you are messing with pointers, don't
    test your program under windows. It will usually appear to be NORMAL.
    Shitty win95.


March 98
--------
1   Change next_pattern() and prev_pattern() to next_mod_track() and
    prev_mod_track() respectively. Fixed a bug in is_mod_playing() and
    resume_mod(). Added goto_mod_track() and finished destroy_mod().
    Change readme.txt and my jgmod homepage contents.

2   Library uploaded. Fixed a pattern break bug. I was not reading the
    documentation properly. Corrected the set pan effect.

3   Redo the s3m loader because I forgot most of the s3m loader codes I did
    2 months ago. 

4   Continue the s3m loader. Able to load 8 bit and 16 bit samples. Only
    tested with 8bit unsigned samples. Fixed a sample offset bug. Tested the
    library with cpp program again. This time it works. This is puzzling.

5   Fixed the 16bit samples loader. Instrument lenght given in S3M is in
    samples. I thought it is in bytes. Change the way notes are stored.
    Previously notes are stored in a single huge array. Now, store notes
    in arrays according to their pattern. I change it because it would be

src/jgmod.h
#ifndef JGMOD_H
#define JGMOD_H


#ifndef JGM_ID
    #define JGM_ID  DAT_ID('J','G','M',' ')
#endif

#ifndef null
#define null    0
#endif

#ifndef uchar
#define uchar   unsigned char
#endif

#ifndef ushort
#define ushort  unsigned short
#endif

#ifndef uint
#define uint    unsigned int
#endif

#ifndef NULL
#define NULL 0
#endif

#ifndef TRUE
#define TRUE -1
#endif

#ifndef FALSE
#define FALSE 0
#endif

#ifdef __cplusplus
extern "C" {
#endif

#define JGMOD_AUTHOR        "Guan Foo Wah"
#define JGMOD_VERSION       0
#define JGMOD_SUB_VERSION   99
#define JGMOD_VERSION_STR   "0.99"
#define JGMOD_DATE_STR      "15 October 2001"
#define JGMOD_DATE          20021015        /* yyyymmdd */

#define NTSC                3579546L
#define JGMOD_PRIORITY      192
#define MAX_ALLEG_VOICE     64
#define LOOP_OFF            0
#define LOOP_ON             1
#define LOOP_BIDI           2

#define ENV_ON              1
#define ENV_SUS             2
#define ENV_LOOP            4

#define XM_MODE             1
#define PERIOD_MODE         2
#define LINEAR_MODE         4


#define PTEFFECT_0          0
#define PTEFFECT_1          1
#define PTEFFECT_2          2
#define PTEFFECT_3          3
#define PTEFFECT_4          4
#define PTEFFECT_5          5
#define PTEFFECT_6          6
#define PTEFFECT_7          7
#define PTEFFECT_8          8
#define PTEFFECT_9          9
#define PTEFFECT_A          10
#define PTEFFECT_B          11
#define PTEFFECT_C          12
#define PTEFFECT_D          13
#define PTEFFECT_E          14
#define PTEFFECT_F          15

#define S3EFFECT_A          16
#define S3EFFECT_D          17
#define S3EFFECT_E          18
#define S3EFFECT_F          19
#define S3EFFECT_I          20
#define S3EFFECT_J          21
#define S3EFFECT_K          22
#define S3EFFECT_L          23
#define S3EFFECT_Q          24
#define S3EFFECT_R          25
#define S3EFFECT_T          26
#define S3EFFECT_U          27
#define S3EFFECT_V          28
#define S3EFFECT_X          29

#define XMEFFECT_1          30
#define XMEFFECT_2          31
#define XMEFFECT_5          32
#define XMEFFECT_6          33
#define XMEFFECT_A          34
#define XMEFFECT_H          35
#define XMEFFECT_K          36
#define XMEFFECT_L          37
#define XMEFFECT_P          38
#define XMEFFECT_X          39


//this is used in get_mod_info() function.
#define MOD15_TYPE          1
#define MOD31_TYPE          2
#define S3M_TYPE            3
#define XM_TYPE             4
#define IT_TYPE             5
#define JGM_TYPE            6
#define UNREAL_S3M_TYPE     7
#define UNREAL_XM_TYPE      8
#define UNREAL_IT_TYPE      9

//-- Header ------------------------------------------------------------------
typedef struct ENVELOPE_INFO
{
    int env[12];
    int pos[12];
    int flg;
    int pts;
    int loopbeg;
    int loopend;
    int susbeg;
    int susend;
    int a;
    int b;
    int p;
    int v;

}ENVELOPE_INFO;


typedef struct CHANNEL_INFO
{
    ENVELOPE_INFO volenv;
    ENVELOPE_INFO panenv;

    int instrument;
    int sample;
    int volume;
    int note;
    int period;
    int c2spd;
    int transpose;
    int pan;
    int kick;       // TRUE if sample needs to be restarted
    int keyon;
    int volfade;    // volume fadeout
    int instfade;   // how much volume to subtract from volfade

    int temp_volume;
    int temp_period;
    int temp_pan;

    int pan_slide_common;
    int pan_slide;
    int pan_slide_left;
    int pan_slide_right;

    int pro_pitch_slide_on;
    int pro_pitch_slide;
    int pro_fine_pitch_slide;
    int s3m_pitch_slide_on;
    int s3m_pitch_slide;
    int s3m_fine_pitch_slide;
    int xm_pitch_slide_up_on;
    int xm_pitch_slide_up;
    int xm_pitch_slide_down_on;
    int xm_pitch_slide_down;
    int xm_fine_pitch_slide_up;
    int xm_fine_pitch_slide_down;
    int xm_extra_fine_pitch_slide_up;
    int xm_x_up;
    int xm_x_down;

    int pro_volume_slide;
    int s3m_volume_slide_on;
    int s3m_fine_volume_slide;
    int s3m_volume_slide;
    int xm_volume_slide_on;
    int xm_volume_slide;
    int xm_fine_volume_slide_up;
    int xm_fine_volume_slide_down;

    int loop_on;
    int loop_times;
    int loop_start;

    int tremolo_on;
    int tremolo_waveform;
    char tremolo_pointer;
    int tremolo_speed;
    int tremolo_depth;
    int tremolo_shift;
    
    int vibrato_on;    
    int vibrato_waveform;
    char vibrato_pointer;
    int vibrato_speed;
    int vibrato_depth;
    int vibrato_shift;

    int slide2period_on;
    int slide2period_spd;
    int slide2period;

    int arpeggio_on;
    int arpeggio;

    int tremor_on;
    int tremor_count;
    int tremor_set;

    int delay_sample;
    int cut_sample;
    int glissando;
    int retrig;
    int s3m_retrig_on;
    int s3m_retrig;
    int s3m_retrig_slide;

    int sample_offset_on;
    int sample_offset;

    int global_volume_slide_on;
    int global_volume_slide;

}CHANNEL_INFO; 

typedef struct MUSIC_INFO
{
    int max_chn;
    int no_chn;

    int tick;
    int pos;
    int pat;
    int trk;
    int flag;

    int bpm;
    int tempo;
    int speed_ratio;
    int pitch_ratio;
    int global_volume;
    
    int new_pos;        // for pattern break
    int new_trk;        // or position jump
    int pattern_delay;  // pattern delay

    int skip_pos;       // for next_pattern
    int skip_trk;       // or prev_pattern
    int loop;           // replay the music if ended
    int pause;          // for pause function
    int forbid;
    int is_playing;

}MUSIC_INFO; 

typedef struct NOTE_INFO
{
    int sample;
    int note;
    int volume;
    int command;
    int extcommand;
}NOTE_INFO;


typedef struct SAMPLE_INFO
{
    int lenght;
    int c2spd;
    int transpose;
    int volume;
    int pan;
    int repoff;
    int replen;
    int loop;

    int vibrato_type;
    int vibrato_spd;
    int vibrato_depth;
    int vibrato_rate;

}SAMPLE_INFO;


typedef struct INSTRUMENT_INFO
{
    int sample_number[96];

    int volenv[12];
    int volpos[12];
    int no_volenv;
    int vol_type;
    int vol_susbeg;
    int vol_susend;
    int vol_begin;
    int vol_end;

    int panenv[12];
    int panpos[12];
    int no_panenv;
    int pan_type;
    int pan_susbeg;
    int pan_susend;
    int pan_begin;
    int pan_end;

    int volume_fadeout;

}INSTRUMENT_INFO;


typedef struct PATTERN_INFO
{
    NOTE_INFO *ni;
    int no_pos;
}PATTERN_INFO;

typedef struct JGMOD
{
    char name[29];
    SAMPLE_INFO *si;
    PATTERN_INFO *pi;
    INSTRUMENT_INFO *ii;
    SAMPLE *s;

    int no_trk;
    int no_pat;
    int pat_table[256];
    int panning[MAX_ALLEG_VOICE];
    int flag;

    int tempo;
    int bpm;

    int restart_pos;       
    int no_chn;
    int no_instrument;
    int no_sample;
    int global_volume;

}JGMOD;

typedef struct JGMOD_INFO
{
    int type;
    char type_name[20];
    char name[29];

}JGMOD_INFO;

//-- externs -----------------------------------------------------------------

extern JGMOD *of;
extern volatile MUSIC_INFO mi;
extern volatile int voice_table[];
extern volatile CHANNEL_INFO ci[MAX_ALLEG_VOICE];
extern volatile int mod_volume;
extern int fast_loading;
extern int enable_m15;
extern int enable_lasttrk_loop;
extern char jgmod_error[];

//-- Prototypes --------------------------------------------------------------
int install_mod(int no_voices);
void remove_mod (void);
JGMOD *load_mod (char *filename);
void mod_interrupt (void);
void play_mod (JGMOD *j, int loop);
void next_mod_track (void);
void prev_mod_track (void);
void goto_mod_track (int new_track);
void stop_mod (void);
int is_mod_playing (void);
void pause_mod (void);
void resume_mod (void);
int is_mod_paused (void);
void destroy_mod (JGMOD *j);
void set_mod_volume (int volume);
int get_mod_volume (void);
SAMPLE *get_jgmod_sample (JGMOD *j, int sample_no);
void set_mod_speed (int speed);
void set_mod_pitch (int pitch);
void toggle_pause_mode (void);
void register_datafile_jgmod(void);
void destroy_mod_datafile(void *j);
int get_mod_info (char *filename, JGMOD_INFO *ji);

#ifdef __cplusplus
}
#endif

#endif  // for JGMOD_H
